File: 637. Average of Levels in Binary Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  vector<double> bfs(TreeNode* start){
    vector<TreeNode*> myQueue {start};
    vector<double> result;
    while (myQueue.size() != 0){
      double total = 0;
      double count = 0;
      vector<TreeNode*> newQueue;
      for (int index = 0; index < myQueue.size(); index++){
        TreeNode* current = myQueue[index];
        if (current != NULL){
          total += current->val;
          count += 1;
        }
        if (current->left != NULL){
          newQueue.push_back(current->left);
        }
        if (current->right != NULL){
          newQueue.push_back(current->right);
        }
      }
      double average = total / count;
      result.push_back(average);
      myQueue = newQueue;
    }
    return result;
  }
public:
  vector<double> averageOfLevels(TreeNode* root) {
    return bfs(root);
  }
};

File: 637. Average of Levels in Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  List<Double> bfs(TreeNode start){
    List<TreeNode> myQueue = new ArrayList<TreeNode>(Arrays.asList(start));
    List<Double> result = new ArrayList<Double>();
    while (myQueue.size() != 0){
      Double total = 0.0;
      Double count = 0.0;
      List<TreeNode> newQueue = new ArrayList<TreeNode>();
      for (int index = 0; index < myQueue.size(); index++){
        TreeNode current = myQueue.get(index);
        if (current != null){
          total += current.val;
          count += 1;
        }
        if (current.left != null){
          newQueue.add(current.left);
        }
        if (current.right != null){
          newQueue.add(current.right);
        }
      }
      Double average = total / count;
      result.add(average);
      myQueue = newQueue;
    }
    return result;
  }
  public List<Double> averageOfLevels(TreeNode root) {
    return bfs(root);
  }
}